src/compiler/checker.ts

```diff
@@ -4981,6 +4991,29 @@ namespace ts {
             return checkTypeRelatedTo(sourceType, targetType, assignableRelation, /*errorNode*/ undefined);
         }
 
+        function isImplementationCompatibleWithOverload(implementation: Signature, overload: Signature): boolean {
+            const erasedSource = getErasedSignature(implementation);
+            const erasedTarget = getErasedSignature(overload);
+
+            // First see if the return types are compatible in either direction.
+            const sourceReturnType = getReturnTypeOfSignature(erasedSource);
+            const targetReturnType = getReturnTypeOfSignature(erasedTarget);
+            if (targetReturnType === voidType
+                || checkTypeRelatedTo(targetReturnType, sourceReturnType, assignableRelation, /*errorNode*/ undefined)
+                || checkTypeRelatedTo(sourceReturnType, targetReturnType, assignableRelation, /*errorNode*/ undefined)) {
+
+                // The return types are compatible, so create versions of the signature with 'any' as the return type.
+                // We need to do this so that we can check assignability while disregarding the return type.
+                const anyReturningSource = getAnyReturningErasedSignature(implementation);
+                const anyReturningTarget = getAnyReturningErasedSignature(overload);
+
+                // Create object types to actually perform relation checks.
+                const anyReturningSourceType = getOrCreateTypeFromSignature(anyReturningSource);
+                const anyReturningTargetType = getOrCreateTypeFromSignature(anyReturningTarget);
+                return checkTypeRelatedTo(anyReturningSourceType, anyReturningTargetType, assignableRelation, /*errorNode*/ undefined);

```

Comment:
We're computing and caching an awful lot of types here. They'll sit in the cache of their associated signature forever and also gum up the assignable relation cache. I think we need a function similar to `compareSignatures` that checks for assignability instead of identity. The function could have an `ignoreReturnTypes` flag and then you would need to create all of these new signatures and types.